src/main/scala/com/microsoft/hyperspace/index/rules/HyperspaceCheck.scala

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.rules
+
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+import com.microsoft.hyperspace.ActiveSparkSession
+import com.microsoft.hyperspace.index.IndexLogEntry
+
+trait HyperspaceCheck {
+  /**
+   * @return Failure reason if an index is excluded in the check.
+   */
+  def reason: String
+}
+
+/**
+ * HyperspaceCheck used in CandidateIndexCollector.
+ */
+trait HyperspaceSourceCheck extends HyperspaceCheck with ActiveSparkSession {
+  def apply(plan: LogicalPlan, indexes: Seq[IndexLogEntry]): Seq[IndexLogEntry]
```

Comment:
Could you add some comments on the return value? Is this also filtering out the given indexes? Then it's doing more than "checking"? Maybe "filter" is a better name?

Reply:
Yes mainly filtering out indexes. But there could be reordering indexes or checking plan (filter node check or join node check - though it also returns empty index list if it fails). Maybe `Filter` is better?